### PR TITLE
#492 Parse duration cap implemented for PDFs

### PR DESCRIPTION
**Pull request by cbeach47**
Initially explored a way to interrupt the indexer thread, but the Tika AutoParser doesn't acknowledge interrupts.  I ended up creating a 'capped' parse handler for the Tika parser so it won't run indefinitely.  

New `com.tle.core.freetext` > `optional.properties`:
```
# Tuning config for attachment indexing.  Currently implemented for PDFs.
# The attachment indexing duration will go over this amount (in milliseconds), 
# but will be checked each time the parser returns data to the controller, 
# so it will be close
textExtracter.parseDurationCap=30000

# Tuning config for attachment indexing.  Specifies how many times to ignore 
# checking that the duration has not gone over it's cap before checking.  Slight 
# performance boost for higher values
textExtracter.durationCheckFrequency=15
```

### CODE REVIEW

**Comment by cbeach47** [comment]
@doolse  - not sure why _codeclimate_ is complaining since I changed the interface `TextExtracterExtension`.   Unless you know of another place _codeclimate_ is keying off of for the method signatures.

Any concerns with merging this pull request?  It'd be nice to get it in before 6.6 is branched.

File: Source/Plugins/Core/com.equella.base/src/com/tle/core/TextExtracterExtension.java [hunk 1]
```diff
@@ -26,7 +26,7 @@
  */
 public interface TextExtracterExtension
 {
-	void extractText(String mimeType, InputStream input, StringBuilder outputText, int maxSize) throws IOException;
+	void extractText(String mimeType, InputStream input, StringBuilder outputText, int maxSize, long parseDurationCap, int durationCheckFrequency) throws IOException;
```
Comment (doolse) [inline Source/Plugins/Core/com.equella.base/src/com/tle/core/TextExtracterExtension.java]: I don't think "durationCheckFrequency" should be a parameter, it's very specific to the technique your using to check the timeout in the CappedBodyContentHolder.

File: Source/Plugins/Core/com.equella.core/src/com/tle/core/freetext/extracter/handler/CappedBodyContentHandler.java [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+
+package com.tle.core.freetext.extracter.handler;
+
+import org.apache.tika.sax.BodyContentHandler;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.xml.sax.ContentHandler;
+import org.xml.sax.SAXException;
+
+public class CappedBodyContentHandler extends BodyContentHandler {
```
Comment (doolse) [inline Source/Plugins/Core/com.equella.core/src/com/tle/core/freetext/extracter/handler/CappedBodyContentHandler.java]: Did you try any other approaches to doing a timeout, such as java.util.concurrent.Future? I'll merge with this implementation anyway as I don't think it'll cause too much of a slow down.

Comment (cbeach47) [reply Source/Plugins/Core/com.equella.core/src/com/tle/core/freetext/extracter/handler/CappedBodyContentHandler.java]: Yes, I tried to interrupt the parser, but from the documentation, the AutoParser doesn't check for those exceptions.

**Comment by cbeach47** [comment]
@doolse - thanks for the review.  I've removed the duration check frequency parameter and hard-coded it to 200.

**Pull request merged**